### PR TITLE
Add argument matcher that ignores the order of elements

### LINKED ISSUE #58: Support order-independent comparison of containers

### PR DESCRIPTION
**Pull request by bash**
Resolves #58

### CODE REVIEW

**Comment by jnferner** [comment]
Why do we need a `Vec` matcher when we have a `Slice` matcher?. I assume you have tried it, just curious about your findings.

**Comment by bash** [comment]
We have a slice matcher?

**Comment by jnferner** [comment]
Sry, meant the `compare_slices_unordered` function

**Comment by bash** [comment]
Exposing `compare_slices_unordered` makes little sense, since it directly compares two slices, instead of returning an `ArgumentMatcher` like `unordered_vec_eq` does.

**Comment by jnferner** [comment]
🤔 Can we expose an `ArgumentMatcher` that works with slices for the sake of generality?

**Comment by bash** [comment]
The `ArgumentMatcher` can compare with a slice:
https://github.com/myelin-ai/mockiato/blob/583976c6fc1e737dbb6b24a1af7703b3c19fb5ea/mockiato/src/internal/matcher/unordered_vec_eq.rs#L39

Do you also want to be able to have a slice as the expected value?

**Comment by bash** [comment]
[ ] Reminder for myself: There needs to be an impl for mutable slices:

```rust
impl<'a, T, U> ArgumentMatcher<&'a mut [U]> for UnorderedVecArgumentMatcher<T> 
```

**Comment by jnferner** [comment]
> Do you also want to be able to have a slice as the expected value?
Precisely, many data structures allow viewing them as a slice, so supporting them instead of relying on a `Vec` allows us to be more generic. Note that we can still store the slice internally as a `Vec`.

**Comment by bash** [comment]
Accepting slices would require all elements to be `Clone`, which is a restriction I'm not happy with.

File: mockiato/src/internal/matcher/unordered_vec_eq.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    fn rejects_slices_with_different_lengths() {
+        assert!(!compare_slices_unordered(&[1], &[1, 2]));
+    }
+
+    #[test]
+    fn works_with_duplicated_items() {
+        assert!(compare_slices_unordered(
+            &["foo", "foo", "bar"],
+            &["foo", "bar"]
+        ))
```
[the remark this answers is not in the document]
Comment (jnferner) [reply mockiato/src/internal/matcher/unordered_vec_eq.rs]: Agreed

[the remark this answers is not in the document]
Comment (bash) [reply mockiato/src/internal/matcher/unordered_vec_eq.rs]: Resolved in e41df81

**Pull request merged**